client/consensus/manual-seal/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+use sr_primitives::traits::{Block as BlockT};
+use sr_primitives::Justification;
+use parking_lot::Mutex;
+use futures::prelude::*;
+use transaction_pool::txpool::{self, Pool as TransactionPool};
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+pub mod rpc;
+use rpc::EngineCommand;
+
+/// The synchronous block-import worker of the engine.
+pub struct ManualSealBlockImport<I> {
+	inner: I,
+}
+
+impl<I> From<I> for ManualSealBlockImport<I> {
+	fn from(i: I) -> Self {
+		ManualSealBlockImport { inner: i }
+	}
+}
+
+impl<B: BlockT, I: BlockImport<B>> BlockImport<B> for ManualSealBlockImport<I> {
+	type Error = I::Error;
+
+	fn check_block(&mut self, block: BlockCheckParams<B>) -> Result<ImportResult, Self::Error>
+	{
+		self.inner.check_block(block)
+	}
+
+	fn import_block(
+		&mut self,
+		block: BlockImportParams<B>,
+		cache: HashMap<CacheKeyId, Vec<u8>>,
+	) -> Result<ImportResult, Self::Error> {
+		// TODO: strip out post-digest.
+
+		self.inner.import_block(block, cache)
+	}
+}
+
+/// The verifier for the manual seal engine; instantly finalizes.
+struct ManualSealVerifier;
+
+impl<B: BlockT> Verifier<B> for ManualSealVerifier {
+	fn verify(
+		&mut self,
+		origin: BlockOrigin,
+		header: B::Header,
+		justification: Option<Justification>,
+		body: Option<Vec<B::Extrinsic>>,
+	) -> Result<(BlockImportParams<B>, Option<Vec<(CacheKeyId, Vec<u8>)>>), String> {
+		let import_params = BlockImportParams {
+			origin,
+			header,
+			justification,
+			post_digests: Vec::new(),
+			body,
+			finalized: true,
+			auxiliary: Vec::new(),
+			fork_choice: ForkChoiceStrategy::LongestChain,
+			allow_missing_state: false,
+		};
+
+		Ok((import_params, None))
+	}
+}
+
+/// Instantiate the import queue for the manual seal consensus engine.
+pub fn import_queue<B: BlockT>(block_import: BoxBlockImport<B>) -> BasicQueue<B>
+{
+	BasicQueue::new(
+		ManualSealVerifier,
+		block_import,
+		None,
+		None,
+	)
+}
+
+/// Creates the background authorship task for the manual seal engine.
+pub async fn run_manual_seal<B, E, A, C, S>(
+	block_import: BoxBlockImport<B>,
+	env: E,
+	pool: Arc<TransactionPool<A>>,
+	seal_block_channel: S,
+	select_chain: C,
+	inherent_data_providers: inherents::InherentDataProviders,
+)
+	where
+		B: BlockT + 'static,
+		E: Environment<B> + 'static,
+		A: txpool::ChainApi + 'static,
+		S: Stream<Item=EngineCommand> + 'static,
+		C: SelectChain<B> + 'static,
+{
+	let block_import = Arc::new(Mutex::new(block_import));
+	let env = Arc::new(Mutex::new(env));
+	let select_chain = Arc::new(select_chain);
+	let inherent_data_providers = Arc::new(inherent_data_providers);
+	let moved_pool = pool.clone();
+
+	seal_block_channel
+		.for_each(move |command| {
+			let select_chain = select_chain.clone();
+			let env = env.clone();
+			let inherent_data_providers = inherent_data_providers.clone();
+			let block_import = block_import.clone();
+			let moved_pool = moved_pool.clone();
+
+			async move {
```

Comment:
Why do we need another closure here? Can't we make the upper one `async`?

Reply:
rust's move semantics, the async block would move the variables and they'd be unavailable for the next call to `for_each`. would work if `for_each` took an `FnOnce`

Reply:
```
.for_each(async move |command| {
            // <cut, since not needed>
			//async move {
           ...
})
```
Wouldn't that work? Why have a regular closure that returns an `async` block if you can have an `async` closure instead?

Reply:
async closures aren't finalized yet